Add sorting bar to Peers page

johnny9 opened the pull request:
[![Windows](https://img.shields.io/badge/OS-Windows-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/268)
[![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/268)
[![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/268)
[![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/insecure_android_apk.zip?branch=pull/268)

johnny9 commented:
Updates from 8ab488f to  c059cf3:
  - Fix spacing between ToggleButtons
  - Fix vertical spacing for ListView items
  - Make "ID" checked by default

johnny9 commented:
| | | | | | |
|--|--|--|--|--|--|
|![Screenshot from 2023-02-18 23-20-50](https://user-images.githubusercontent.com/985648/219922144-48650549-8a86-4e2a-9ef7-53c0adbfebad.png)|![Screenshot from 2023-02-18 23-21-06](https://user-images.githubusercontent.com/985648/219922146-0a5d523a-f3d0-4a35-b40a-343d0c27358c.png)|![Screenshot from 2023-02-18 23-21-14](https://user-images.githubusercontent.com/985648/219922150-46ab2627-74aa-4c9d-93a3-044d2d714f4b.png)|![Screenshot from 2023-02-18 23-21-21](https://user-images.githubusercontent.com/985648/219922151-682b92e2-a087-433c-8563-b16b14c23d90.png)|![Screenshot from 2023-02-18 23-21-28](https://user-images.githubusercontent.com/985648/219922152-37f4a04d-1eb9-42e8-9383-a89227af73d2.png)|![Screenshot from 2023-02-18 23-21-35](https://user-images.githubusercontent.com/985648/219922157-88daa908-cc21-445e-8a29-bc8f7a4509bd.png)|

jarolrod reviewed (commented):
toggle button text color doesn't work on light mode

<img width="752" alt="Screen Shot 2023-02-20 at 6 51 59 PM" src="https://user-images.githubusercontent.com/23396902/220214503-1dcacd00-dbde-4921-94d2-78938463fec0.png">

jarolrod commented on src/qml/controls/ToggleButton.qml:
```diff
@@ -0,0 +1,68 @@
+        id: buttonText
+        text: parent.text
+        font: parent.font
+        color: Theme.color.white
```
This should actually be
```suggestion
        color: root.textColor
```

jarolrod replied:
This is what was causing [this issue](https://github.com/bitcoin-core/gui-qml/pull/268#pullrequestreview-1306436968)

jarolrod commented on src/qml/controls/ToggleButton.qml:
```diff
@@ -0,0 +1,68 @@
+    property color textActiveColor: Theme.color.neutral9
+    id: root
+    checkable: true
+    hoverEnabled: true
```
This is true by default for `Button`
```suggestion
```

johnny9 replied:
I noticed this on your commit.  It is not on by default in my Ubuntu or Fedora machines.

jarolrod commented on src/qml/controls/ToggleButton.qml:
```diff
@@ -0,0 +1,68 @@
+    topPadding: 5
+    bottomPadding: 5
+
+    contentItem: Text {
```
We could just make this a Header and we can remove specifying font in lines 18-20, and this becomes:

```qml
    contentItem: Header {
        id: buttonText
        header: root.text
        headerSize: 13
        headerColor: root.textColor
    }
```

johnny9 replied:
Does it make sense to just have a "CoreText" item?

jarolrod commented on src/qml/controls/ToggleButton.qml:
```diff
@@ -0,0 +1,68 @@
+import QtQuick.Controls 2.15
+
+Button {
+    property color defaultColor: Theme.color.neutral2
```
i would add a `bg` prefix to these names just to be clear
```suggestion
    property color bgDefaultColor: Theme.color.neutral2
```

jarolrod commented on src/qml/controls/ToggleButton.qml:
```diff
@@ -0,0 +1,68 @@
+        },
+        State {
+            name: "CHECKED"
+            when: root.checked
```
I would inline these next to name, but that's just stylistic

jarolrod commented on src/qml/pages/node/Peers.qml:
```diff
@@ -31,36 +30,89 @@ Page {
         horizontalAlignment: Text.AlignHCenter
     }
 
+    RowLayout {
```
Maybe make this [Flow](https://doc.qt.io/qt-6/qml-qtquick-flow.html) so that the buttons don't go out of view on narrow screens?

| PR | Patch |
| -- | ----- |
| ![2023-02-21_15-05](https://user-images.githubusercontent.com/23396902/220451122-e0dcf0e4-60cc-4b90-aca2-9597c95dc4ae.png) | ![2023-02-21_15-29](https://user-images.githubusercontent.com/23396902/220451421-3e607277-9d5b-40bc-a8c1-b1ab5776118f.png) |

If we want the buttons to always be centered within the width of the flow, we can add some additional logic to the left and right margins




```suggestion
    Flow {
```

johnny9 replied:
Neat QML type/item.  I like it.

jarolrod reviewed (commented):
Concept ACK

Diff on `ToggleButton.qml` 
```diff
diff --git a/src/qml/controls/ToggleButton.qml b/src/qml/controls/ToggleButton.qml
index 05cf54e6a..ea72c5802 100644
--- a/src/qml/controls/ToggleButton.qml
+++ b/src/qml/controls/ToggleButton.qml
@@ -6,38 +6,30 @@ import QtQuick 2.15
 import QtQuick.Controls 2.15

 Button {
-    property color defaultColor: Theme.color.neutral2
-    property color hoverColor: Theme.color.neutral2
-    property color activeColor: Theme.color.neutral5
+    id: root
+    property color bgDefaultColor: Theme.color.neutral2
+    property color bgHoverColor: Theme.color.neutral2
+    property color bgActiveColor: Theme.color.neutral5
     property color textColor: Theme.color.neutral7
     property color textHoverColor: Theme.color.orangeLight1
     property color textActiveColor: Theme.color.neutral9
-    id: root
-    checkable: true
     hoverEnabled: true
-    font.family: "Inter"
-    font.styleName: "Regular"
-    font.pixelSize: 13
+    checkable: true
     leftPadding: 12
     rightPadding: 12
     topPadding: 5
     bottomPadding: 5

-    contentItem: Text {
+    contentItem: Header {
         id: buttonText
-        text: parent.text
-        font: parent.font
-        color: Theme.color.white
-        horizontalAlignment: Text.AlignHCenter
-        verticalAlignment: Text.AlignVCenter
-
-        Behavior on color {
-            ColorAnimation { duration: 150 }
-        }
+        header: root.text
+        headerSize: 13
+        headerColor: root.textColor
     }
+
     background: Rectangle {
         id: bg
-        color: root.defaultColor
+        color: root.bgDefaultColor
         radius: 5

         Behavior on color {
@@ -45,24 +37,22 @@ Button {
         }
     }

-    state:"DEFAULT"
+    state: "DEFAULT"

     states: [
         State {
             name: "DEFAULT"
-            PropertyChanges { target: bg; color: root.defaultColor }
+            PropertyChanges { target: bg; color: root.bgDefaultColor }
         },
         State {
-            name: "CHECKED"
-            when: root.checked
-            PropertyChanges { target: bg; color: root.activeColor }
-            PropertyChanges { target: buttonText; color: root.textActiveColor }
+            name: "CHECKED"; when: root.checked
+            PropertyChanges { target: bg; color: root.bgActiveColor }
+            PropertyChanges { target: buttonText; headerColor: root.textActiveColor }
         },
         State {
-            name: "HOVER"
-            when: root.hovered
-            PropertyChanges { target: bg; color: root.hoverColor }
-            PropertyChanges { target: buttonText; color: root.textHoverColor }
+            name: "HOVER"; when: root.hovered
+            PropertyChanges { target: bg; color: root.bgHoverColor }
+            PropertyChanges { target: buttonText; headerColor: root.textHoverColor }
         }
     ]
 }
```

diff on `Peers.qml`

```diff
diff --git a/src/qml/pages/node/Peers.qml b/src/qml/pages/node/Peers.qml
index 2d9b8b2af..546bbc6b2 100644
--- a/src/qml/pages/node/Peers.qml
+++ b/src/qml/pages/node/Peers.qml
@@ -30,11 +30,12 @@ Page {
         horizontalAlignment: Text.AlignHCenter
     }

-    RowLayout {
+    Flow {
         id: sortSelection
         anchors.top: description.bottom
         anchors.topMargin: 20
         anchors.horizontalCenter: parent.horizontalCenter
+        width: Math.min(parent.width - 40, 450)
         spacing: 10
         ToggleButton {
             text: qsTr("ID")
```

johnny9 commented:
from[ c059cf3 ](https://github.com/bitcoin-core/gui-qml/commit/c059cf3d4ffbc1d5ade23b691e385485d290972d) to   b36b8d5 :
  - fixed togglebutton text color in light mode
  - used CoreText to reduce duplication of Text font declarations
  - removed explicit DEFAULT state
  - prefix togglebutton background color properties with bg
  - inline `when` statements
  - Use Flow to wrap toggle button row

johnny9 commented:
![Screenshot from 2023-02-21 23-44-00](https://user-images.githubusercontent.com/985648/220524138-14f4d9a7-c95b-4ad4-94f2-3647f1c89c73.png)

jarolrod reviewed (commented):
The CoreText component needs review on its own. Can you move these changes out to its own independent PR?
<img width="752" alt="Screen Shot 2023-02-22 at 3 08 52 AM" src="https://user-images.githubusercontent.com/23396902/220560422-381ed75f-a45f-4836-850a-dbecb6cd6844.png">

hebasto commented:
Rebase?

johnny9 commented:
> The CoreText component needs review on its own. Can you move these changes out to its own independent PR?
> PR 	master
> <img alt="Screen Shot 2023-02-22 at 3 08 52 AM" width="752" src="https://user-images.githubusercontent.com/23396902/220560422-381ed75f-a45f-4836-850a-dbecb6cd6844.png"> 	<img alt="Screen Shot 2023-02-22 at 3 10 08 AM" width="752" src="https://user-images.githubusercontent.com/23396902/220560688-adf70ac8-a8d0-43ac-9c6d-3eb071ab0ba2.png">

#274 
Interesting that its wrapping for you. I'm not seeing that behavior. Ill try on the platforms I have.

hebasto commented:
Needs rebase again :)

johnny9 commented:
Update from[ 51a3e21 ](https://github.com/bitcoin-core/gui-qml/commit/51a3e219ecaba975731f753a99855e05508a058b)to a46d12d:
  - rebased with main

jarolrod reviewed (commented):
ACK a46d12d6f37220dc42c6ebc97e4b1be4461c79fc

One thing to note is I believe that the toggle buttons have three states, selected, selected again, not selected. When selected again it should reverse the sorting order; so if selected is least to greatest, selected again would be greatest to least. We can connect with christoph about this soon.
